internal/middlewares/access_manager.go

```diff
@@ -0,0 +1,195 @@
+			IsEmailConfirmed:        !role.IsEmailConfirmed,
+			IsPersonalDataConfirmed: !role.IsPersonalDataConfirmed,
+		})
+	}
+	return allRoleVariations
+}
+
+type path struct {
+	asString string
+	asRegexp *regexp.Regexp
+}
+
+func (p *path) isMatch(st string) bool {
+	endpointPath := normalizePath(st)
+	if p.asString != "" {
+		return p.asString == endpointPath
+	}
+	return p.asRegexp.MatchString(endpointPath)
+}
+
+func normalizePath(endpointPath string) string {
+	if !strings.HasPrefix(endpointPath, "/") {
+		endpointPath = "/" + endpointPath
+	}
+	endpointPath = strings.TrimSuffix(endpointPath, "/")
+	return endpointPath
+}
+
+func newPath(endpointPath string) path {
+	if strings.Contains(endpointPath, "{") {
+		first := strings.Index(endpointPath, "{")
+		second := strings.Index(endpointPath, "}")
+		endpointPath = endpointPath[:first] + "(.*)" + endpointPath[second+1:]
+		return path{
+			asRegexp: regexp.MustCompile(endpointPath),
+		}
+	} else {
+		return path{
+			asString: endpointPath,
+		}
+	}
+}
+
+func endpointConversion(path string) string {
+	return apiPrefix + path
+}
+
+// AddNewAccess adds new access to the access manager. Returns true if access was added, false if access was not added
+func (a *blackListAccessManager) AddNewAccess(role Role, endpointMethod, endpointPath string) (bool, error) {
+	if !utils.IsValueInList(role, a.acceptableRoles) {
+		return false, errors.New(fmt.Sprintf("role %v is not in the list of acceptable roles", role))
+	}
+	if utils.IsValueInList(role, a.fullAccessRoles) {
+		return false, nil
+	}
+	endpointMethod = strings.ToUpper(endpointMethod)
+	paths, ok := a.endpoints[endpointMethod]
+	if ok {
+		if !utils.IsValueInList(endpointPath, paths) {
+			return false, errors.New(fmt.Sprintf("path %s is not in the list of existing endpoints", endpointPath))
+		}
+
+		endpointsByRole, ok2 := a.accessMap[role]
+		if !ok2 {
+			endpointsByRole = make(map[string][]path)
+			a.accessMap[role] = endpointsByRole
+		}
+		pathToUpdate, ok2 := endpointsByRole[endpointMethod]
+		if !ok2 {
+			pathToUpdate = []path{
+				newPath(endpointConversion(endpointPath)),
+			}
+			endpointsByRole[endpointMethod] = pathToUpdate
+			a.accessMap[role] = endpointsByRole
+			return true, nil
+		}
+		if !utils.IsValueInList(newPath(endpointPath), pathToUpdate) {
+			pathToUpdate = append(pathToUpdate, newPath(endpointConversion(endpointPath)))
+			endpointsByRole[endpointMethod] = pathToUpdate
+			a.accessMap[role] = endpointsByRole
+			return true, nil
+		}
+		return false, nil
+
+	}
+	return false, errors.New(fmt.Sprintf("method %s is not in the list of supported endpoints", endpointMethod))
+}
+
+// HasAccess checks if role has access to the endpoint
+func (a *blackListAccessManager) HasAccess(role Role, method, path string) bool {
+	if utils.IsValueInList(role, a.fullAccessRoles) {
+		return true
+	}
+	paths, ok := a.accessMap[role][method]
+	if ok {
+		for _, p := range paths {
+			if p.isMatch(path) {
```

Comment:
To improve the readability of the loop, consider renaming the loop variable to "allowedPath". This will make the code easier to understand and follow.

Reply:
Done

Reply:
Sorry for the misunderstanding. What I meant was to rename the loop variable "p" to something more descriptive that makes the boolean expression easier to understand, such as allowedPath.isMatch(path), instead of using p.isMatch(path).

Reply:
Done